### ISSUE TITLE
"publish --force" and "unpublish VERSION" seem to be broken

### ISSUE THREAD

**Issue by liarco** [verdaccio/verdaccio#1435]
### Describe the bug
I tried to look at #1359 and #1405 but couldn't find a solution.
I'm using ~~https://github.com/Remitly/verdaccio-s3-storage~~ https://registry.verdaccio.org/-/web/detail/verdaccio-aws-s3-storage as a storage plugin, but you can comment the whole `storage` key in configuration and it will be the same _(except for one thing that is pointed out below)_.

#### npm publish --force
If I try to overwrite an existing package version with a new one:
```bash
npm publish --force --registry MY_REPO
```
The result is:

- old package is removed from the remote storage and the web UI doesn't show it
- the `npm` command fails with:
```bash
npm ERR! code EPUBLISHCONFLICT
npm ERR! publish fail Cannot publish over existing version.
npm ERR! publish fail Update the 'version' field in package.json and try again.
npm ERR! publish fail 
npm ERR! publish fail To automatically increment version numbers, see:
npm ERR! publish fail     npm help version

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/marco/.npm/_logs/2019-08-10T13_46_04_312Z-debug.log
```
- running the same exact command again **publishes successfully**

#### npm unpublish --force SPECIFIC_VERSION
If I try to unpublish the package first to avoid errors:
```bash
npm unpublish "@test/my-pack@1.0.0-beta" --force --registry MY_REPO
```

The result is:

- `npm` exits with no errors and says that the package has been removed
- the package still shows up on the AWS S3 storage and on the web UI
- trying to publish the same package from now on will result in failures **(this doesn't happen when using local storage: publishing again works as with the command above)**

The only way I found to recover from this situation was:

- removing the `.tgz` archive manually from the storage
- restart the verdaccio server

These commands don't work as expected.

### To Reproduce
I'm using the following `Dockerfile` to speed up testing:
```Dockerfile
FROM verdaccio/verdaccio:latest

USER root

ENV NODE_ENV=production

# Install "envsubst" to replace ENV_VARS in config
RUN apk add gettext libintl

# Install AWS S3 Plugin
RUN yarn add verdaccio-aws-s3-storage --production=true --no-lockfile --registry https://registry.verdaccio.org

# Copy configuration and .htpasswd
COPY ./config.yml  $VERDACCIO_APPDIR/conf/config.yaml

# Fix permissions
RUN chown -R $VERDACCIO_USER_UID:root $VERDACCIO_APPDIR/conf/config.yaml \
    && chmod -R g=u $VERDACCIO_APPDIR/conf/config.yaml

USER $VERDACCIO_USER_UID

# Set user credentials, replace ENV_VARS in config and finally run Verdaccio
CMD echo "$HTPASSWD" > /verdaccio/storage/htpasswd \
    && envsubst \$AWS_S3_BUCKET,\$AWS_S3_PATH_PREFIX,\$AWS_DEFAULT_REGION,\$AWS_S3_ENDPOINT,\$AWS_S3_PATH_STYLE,\$AWS_ACCESS_KEY_ID,\$AWS_SECRET_ACCESS_KEY < $VERDACCIO_APPDIR/conf/config.yaml > $VERDACCIO_APPDIR/conf/config.yaml \
    && $VERDACCIO_APPDIR/bin/verdaccio --config $VERDACCIO_APPDIR/conf/config.yaml --listen $VERDACCIO_PROTOCOL://0.0.0.0:$VERDACCIO_PORT
```

This `docker-compose.yml` should run your registry:
```yaml
version: '2.1'

services:
  verdaccio:
    image: testing/private-verdaccio-s3:latest
    environment:
      AWS_S3_BUCKET: 
      AWS_S3_PATH_PREFIX: 
      AWS_ACCESS_KEY_ID: 
      AWS_SECRET_ACCESS_KEY: 
      AWS_DEFAULT_REGION:
      AWS_S3_ENDPOINT: 
      AWS_S3_PATH_STYLE: "true"
      # User is demo:demo
      HTPASSWD: |
        demo:$apr1$uq2Uw53i$dS7GoGvNuupumWvXjOJ.R1
    ports:
      - "4874:4873"
```

Configuration file `config.yml`:
```yaml
#
# This is the config file used for the docker images.
# It allows all users to do anything, so don't use it on production systems.
#
# Do not configure host and port under `listen` in this file
# as it will be ignored when using docker.
# see https://verdaccio.org/docs/en/docker#docker-and-custom-port-configuration
#
# Look here for more config file examples:
# https://github.com/verdaccio/verdaccio/tree/master/conf
#

# path to a directory with all packages
storage: /verdaccio/storage/data

# comment this section entirely to disable AWS S3
store:
  aws-s3-storage:
    bucket: $AWS_S3_BUCKET
    keyPrefix: $AWS_S3_PATH_PREFIX
    region: $AWS_DEFAULT_REGION
    endpoint: $AWS_S3_ENDPOINT
    s3ForcePathStyle: $AWS_S3_PATH_STYLE
    accessKeyId: $AWS_ACCESS_KEY_ID
    secretAccessKey: $AWS_SECRET_ACCESS_KEY

web:
  title: Verdaccio

auth:
  htpasswd:
    file: /verdaccio/storage/htpasswd
    max_users: -1

security:
  api:
    jwt:
      sign:
        expiresIn: 60d
        notBefore: 1
  web:
    sign:
      expiresIn: 7d
      notBefore: 1

# a list of other known repositories we can talk to
uplinks:
  npmjs:
    url: https://registry.npmjs.org/

packages:
  '**':
    access: $all
    publish: $all
    unpublish: $all

    # if package is not available locally, proxy requests to 'npmjs' registry
    proxy: npmjs

middlewares:
  audit:
    enabled: true

# log settings
logs:
  - { type: stdout, format: pretty, level: http }

listen:
  - 0.0.0.0:4873
```

Put these 3 files in a folder called `verdaccio-test`, then build and run everything:
```bash
cd /path/to/verdaccio-test
docker build -t testing/private-verdaccio-s3:latest .
docker-compose up -d

# Attach to logs
docker logs -f verdaccio-test_verdaccio_1

# use http://127.0.0.1:4874
```

### Expected behavior
Packages should be deleted correctly.

### Conclusion
I hope that everything is clear enough, please feel free to ask for further information.

Thank you for your time.

**Comment by juanpicado** [comment]
Did you see the lates release ?  Previous version of verdaccio 4 might have corrupted the metadata. Try again with the lates version or you might need to update it manually.  The problem is fixed in 4.2.1

**Comment by liarco** [comment]
I'm sorry that I didn't specify this before, but I think that I'm already using `4.2.1` (at least I see `Powered by Verdaccio / 4.2.1` in the footer).

> One thing to be clear, the latest update might be problematic since does not autofix the corrupted metadata. I can explain how to do it manually is really straightforward.

Thank you for pointing this out, but I'm just starting using Verdaccio, so I don't have to migrate anything.

**Comment by juanpicado** [comment]
Then if you are starting form scratch the easier is remove the package corrupted (whole folder) and try again.

**Comment by liarco** [comment]
I'm sorry, maybe I'm missing something.

Here is what's happening:

- I start a new registry from scratch (brand new, using latest available version `4.2.1`)
- add a package for testing purpose
- run the commands above
- get the unexpected behavior

I don't understand why this thing is happening since I'm using the fixed version from the beginning...

I plan to use this feature in order to release `-dev` versions containing the latest available commits, so I will need to republish packages many times.

**Comment by juanpicado** [comment]
Have you tried with a empty and clean storage ?

**Comment by liarco** [comment]
Yep, the purpose of commenting the whole AWS S3 configuration was to try everything with a clean (and self-contained) registry.

**Comment by juanpicado** [comment]
Could you share a repo with example of your configuration + docker etc ? I would like to try locally.

**Comment by liarco** [comment]
Absolutely! The files are exactly the ones that I shared above, but I'm gonna create a demo repo in a matter of minutes. Thank you.

In the mean while I tested one more time to be sure and it's still the same:

My console:
```bash
MacBook-Air-di-Marco:test marco$ npm publish --force --registry http://127.0.0.1:4874
npm WARN using --force I sure hope you know what you are doing.
npm notice 
npm notice 📦  @test/beta-pack@0.0.0-dev
npm notice === Tarball Contents === 
npm notice 49B  index.js    
npm notice 229B package.json
npm notice 20B  README.md   
npm notice === Tarball Details === 
npm notice name:          @test/beta-pack                         
npm notice version:       0.0.0-dev                               
npm notice package size:  373 B                                   
npm notice unpacked size: 298 B                                   
npm notice shasum:        0999bc22c3e87051fa3bc945ca39cc276e9b0b3b
npm notice integrity:     sha512-G+b1QLXVycRYC[...]KEUVGCUkov2GQ==
npm notice total files:   3                                       
npm notice 
+ @test/beta-pack@0.0.0-dev
MacBook-Air-di-Marco:test marco$ npm publish --force --registry http://127.0.0.1:4874
npm WARN using --force I sure hope you know what you are doing.
npm notice 
npm notice 📦  @test/beta-pack@0.0.0-dev
npm notice === Tarball Contents === 
npm notice 49B  index.js    
npm notice 229B package.json
npm notice 20B  README.md   
npm notice === Tarball Details === 
npm notice name:          @test/beta-pack                         
npm notice version:       0.0.0-dev                               
npm notice package size:  373 B                                   
npm notice unpacked size: 298 B                                   
npm notice shasum:        0999bc22c3e87051fa3bc945ca39cc276e9b0b3b
npm notice integrity:     sha512-G+b1QLXVycRYC[...]KEUVGCUkov2GQ==
npm notice total files:   3                                       
npm notice 
npm WARN publish Forced publish over @test/beta-pack@0.0.0-dev
npm ERR! code EPUBLISHCONFLICT
npm ERR! publish fail Cannot publish over existing version.
npm ERR! publish fail Update the 'version' field in package.json and try again.
npm ERR! publish fail 
npm ERR! publish fail To automatically increment version numbers, see:
npm ERR! publish fail     npm help version

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/marco/.npm/_logs/2019-08-10T15_21_47_692Z-debug.log
MacBook-Air-di-Marco:test marco$ npm publish --force --registry http://127.0.0.1:4874
npm WARN using --force I sure hope you know what you are doing.
npm notice 
npm notice 📦  @test/beta-pack@0.0.0-dev
npm notice === Tarball Contents === 
npm notice 49B  index.js    
npm notice 229B package.json
npm notice 20B  README.md   
npm notice === Tarball Details === 
npm notice name:          @test/beta-pack                         
npm notice version:       0.0.0-dev                               
npm notice package size:  373 B                                   
npm notice unpacked size: 298 B                                   
npm notice shasum:        0999bc22c3e87051fa3bc945ca39cc276e9b0b3b
npm notice integrity:     sha512-G+b1QLXVycRYC[...]KEUVGCUkov2GQ==
npm notice total files:   3                                       
npm notice 
+ @test/beta-pack@0.0.0-dev
MacBook-Air-di-Marco:test marco$ 
```

Container logs:
```bash
 http <-- 201, user: liarco(192.168.160.1), req: 'PUT /@test%2fbeta-pack', bytes: 1380/53
 http <-- 409, user: liarco(192.168.160.1), req: 'PUT /@test%2fbeta-pack', error: this package is already present
 http <-- 200, user: liarco(192.168.160.1), req: 'GET /@test%2fbeta-pack?write=true', bytes: 0/592
 http <-- 304, user: liarco(192.168.160.1), req: 'GET /@test%2fbeta-pack?write=true', bytes: 0/0
 http <-- 201, user: liarco(192.168.160.1), req: 'DELETE /@test%2fbeta-pack/-rev/3-ef8a0450baebcd97', bytes: 0/30
 http <-- 201, user: liarco(192.168.160.1), req: 'PUT /@test%2fbeta-pack', bytes: 1380/53
```

**Comment by liarco** [comment]
@juanpicado here it is: https://github.com/liarco/verdaccio-troubleshooting

**Comment by juanpicado** [comment]
One thing, you are using `verdaccio-aws-s3-storage` according your description says https://github.com/Remitly/verdaccio-s3-storage.

Anyway, both are almost the same, but the `verdaccio-aws-s3-storage` is under my control and has some bugfixes.

I could not reproduce your issue with https://github.com/verdaccio/docker-examples/tree/master/amazon-s3-docker-example/v4

which is almost the same. 

Are you using a new clean bucket in S3 ?

**Comment by juanpicado** [comment]
If you change the logger to `trace`, you will have crazy useful logging.

**Comment by liarco** [comment]
> One thing, you are using `verdaccio-aws-s3-storage` according your description says https://github.com/Remitly/verdaccio-s3-storage.
> 
> Anyway, both are almost the same, but the `verdaccio-aws-s3-storage` is under my control and has some bugfixes.

You are right, sorry for that, I picked the wrong repo since I tried both. 🤦‍♂ 

> I could not reproduce your issue with https://github.com/verdaccio/docker-examples/tree/master/amazon-s3-docker-example/v4
> 
> which is almost the same.
> 
> Are you using a new clean bucket in S3 ?

Yes, I wiped it out completely several times... the strange thing is that it happens to me even if I run it **without** S3. The example repo that I shared is configured to skip S3 exactly to take it off the equation.

So, running my example, you are not getting the same result as shown in the `README.md` file? 😢

**Comment by liarco** [comment]
Here are my logs with `trace` level. I don't understand why I get the `error: this package is already present`, but then package is deleted and not republished _**(second try)**_.

**Important:** I'm not using S3 during this test, it's just local storage inside the container without any persistence.

**First try:** _(published fine)_
```bash
verdaccio_1  |  info <-- 192.168.192.1 requested 'PUT /@test%2fmy-pack'
verdaccio_1  |  trace--- api middleware using JWT auth token
verdaccio_1  |  trace--- [middleware/allow][publish] allow for liarco
verdaccio_1  |  trace--- allow publish for @test/my-pack init | plugins: 2
verdaccio_1  |  trace--- allow publish for @test/my-pack plugin does not implement allow_publish
verdaccio_1  |  trace--- [auth/allow_action]: user: undefined
verdaccio_1  |  trace--- [auth/allow_action]: hasPermission? true for user: undefined
verdaccio_1  |  trace--- auth/allow_action: access granted to: undefined
verdaccio_1  |  trace--- allowed publish for @test/my-pack
verdaccio_1  |  debug--- publishing or updating a new version for @test/my-pack
verdaccio_1  |  debug--- adding a new version for @test/my-pack
verdaccio_1  |  http <-- 201, user: liarco(192.168.192.1), req: 'PUT /@test%2fmy-pack', bytes: 1128/53
```

**Second try:** _(`publish` fails, but package has been removed)_
```bash
verdaccio_1  |  info <-- 192.168.192.1 requested 'PUT /@test%2fmy-pack'
verdaccio_1  |  trace--- api middleware using JWT auth token
verdaccio_1  |  trace--- [middleware/allow][publish] allow for liarco
verdaccio_1  |  trace--- allow publish for @test/my-pack init | plugins: 2
verdaccio_1  |  trace--- allow publish for @test/my-pack plugin does not implement allow_publish
verdaccio_1  |  trace--- [auth/allow_action]: user: undefined
verdaccio_1  |  trace--- [auth/allow_action]: hasPermission? true for user: undefined
verdaccio_1  |  trace--- auth/allow_action: access granted to: undefined
verdaccio_1  |  trace--- allowed publish for @test/my-pack
verdaccio_1  |  debug--- publishing or updating a new version for @test/my-pack
verdaccio_1  |  debug--- adding a new version for @test/my-pack
verdaccio_1  |  http <-- 409, user: liarco(192.168.192.1), req: 'PUT /@test%2fmy-pack', error: this package is already present
verdaccio_1  |  info <-- 192.168.192.1 requested 'GET /@test%2fmy-pack?write=true'
verdaccio_1  |  trace--- api middleware using JWT auth token
verdaccio_1  |  trace--- [middleware/allow][access] allow for liarco
verdaccio_1  |  trace--- allow access for @test/my-pack
verdaccio_1  |  trace--- [auth/allow_action]: user: undefined
verdaccio_1  |  trace--- [auth/allow_action]: hasPermission? true for user: undefined
verdaccio_1  |  trace--- auth/allow_action: access granted to: undefined
verdaccio_1  |  trace--- allowed access for @test/my-pack
verdaccio_1  |  http <-- 200, user: liarco(192.168.192.1), req: 'GET /@test%2fmy-pack?write=true', bytes: 0/988
verdaccio_1  |  info <-- 192.168.192.1 requested 'GET /@test%2fmy-pack?write=true'
verdaccio_1  |  trace--- api middleware using JWT auth token
verdaccio_1  |  trace--- [middleware/allow][access] allow for liarco
verdaccio_1  |  trace--- allow access for @test/my-pack
verdaccio_1  |  trace--- [auth/allow_action]: user: undefined
verdaccio_1  |  trace--- [auth/allow_action]: hasPermission? true for user: undefined
verdaccio_1  |  trace--- auth/allow_action: access granted to: undefined
verdaccio_1  |  trace--- allowed access for @test/my-pack
verdaccio_1  |  http <-- 304, user: liarco(192.168.192.1), req: 'GET /@test%2fmy-pack?write=true', bytes: 0/0
verdaccio_1  |  info <-- 192.168.192.1 requested 'DELETE /@test%2fmy-pack/-rev/3-1803f2c16d668c6f'
verdaccio_1  |  trace--- api middleware using JWT auth token
verdaccio_1  |  trace--- [middleware/allow][unpublish] allow for liarco
verdaccio_1  |  trace--- allow unpublish for @test/my-pack
verdaccio_1  |  trace--- allow unpublish for @test/my-pack plugin does not implement allow_unpublish
verdaccio_1  |  trace--- fallback unpublish for @test/my-pack has groups: true for liarco
verdaccio_1  |  trace--- allow_action for unpublish for @test/my-pack has groups: true for liarco
verdaccio_1  |  trace--- [auth/allow_action]: user: undefined
verdaccio_1  |  trace--- [auth/allow_action]: hasPermission? true for user: undefined
verdaccio_1  |  trace--- auth/allow_action: access granted to: undefined
verdaccio_1  |  trace--- allowed unpublish for @test/my-pack
verdaccio_1  |  debug--- unpublishing @test/my-pack
verdaccio_1  |  debug-=- [storage] removing package @test/my-pack
verdaccio_1  |  debug-=- [storage/_deleteAttachments] delete attachments total: 1
verdaccio_1  |  http <-- 201, user: liarco(192.168.192.1), req: 'DELETE /@test%2fmy-pack/-rev/3-1803f2c16d668c6f', bytes: 0/30
```

**Third try:** _(works fine since the package is not there)_
```bash
verdaccio_1  |  info <-- 192.168.192.1 requested 'PUT /@test%2fmy-pack'
verdaccio_1  |  trace--- api middleware using JWT auth token
verdaccio_1  |  trace--- [middleware/allow][publish] allow for liarco
verdaccio_1  |  trace--- allow publish for @test/my-pack init | plugins: 2
verdaccio_1  |  trace--- allow publish for @test/my-pack plugin does not implement allow_publish
verdaccio_1  |  trace--- [auth/allow_action]: user: undefined
verdaccio_1  |  trace--- [auth/allow_action]: hasPermission? true for user: undefined
verdaccio_1  |  trace--- auth/allow_action: access granted to: undefined
verdaccio_1  |  trace--- allowed publish for @test/my-pack
verdaccio_1  |  debug--- publishing or updating a new version for @test/my-pack
verdaccio_1  |  debug--- adding a new version for @test/my-pack
verdaccio_1  |  http <-- 201, user: liarco(192.168.192.1), req: 'PUT /@test%2fmy-pack', bytes: 1128/53
```

**Comment by juanpicado** [comment]
Could you do something for me ? Before second step, could you verify whether in the metadata the version you just remove it still in the metadata ? 

Also I would wait feebdack from the other tickets or you can ask them.

The thing is it works for me and the unit testing is green (and believe me I invested a lot of time in the unpublish part) . It is odd does not works for you.

**Comment by liarco** [comment]
First of all, thank you very much for your time. I really appreciate your help with this issue.

Later today I'm gonna send you feedback about what you asked plus some updates that I got after further testing.

**Comment by liarco** [comment]
@juanpicado here are my updates _(with a complete test)_:

### 1. First I publish the package in an empty registry
- `/verdaccio_s3_db.json` is updated correctly
```json
{"list":["@test/demo"],"secret":"4d46495e75b0dd52555a620e08a4f46c9b72f66a77b44c32477d08d9a08eacc1"}
```
- `/@test/demo/package.json` seems to be ok too
```json
{
  "name": "@test/demo",
  "versions": {
    "0.0.0-dev": {
      "name": "@test/demo",
      "version": "0.0.0-dev",
      "author": {
        "name": "AUTHOR_NAME",
        "email": "EMAIL_ADDRESS"
      },
      "main": "index.js",
      "license": "MIT",
      "readmeFilename": "README.md",
      "description": "A simple package to test the registry.",
      "_id": "@test/demo@0.0.0-dev",
      "_nodeVersion": "12.8.0",
      "_npmVersion": "6.10.3",
      "dist": {
        "integrity": "sha512-tfC1KvPmzlyNfdrVHs4a705DY6zHcTj5zY5iz4dDv0ZTtaNF1mt0caC+9DXSrqPLDZ1uUVJUzk0VIfzGlA+s2A==",
        "shasum": "45514c40a481596880710f08ff402682d3d6028f",
        "tarball": "http://MY_REGISTRY_DOMAIN/@test/demo/-/@test/demo-0.0.0-dev.tgz"
      },
      "contributors": []
    }
  },
  "time": {
    "modified": "2019-08-12T14:39:01.177Z",
    "created": "2019-08-12T14:39:01.177Z",
    "0.0.0-dev": "2019-08-12T14:39:01.177Z"
  },
  "users": {},
  "dist-tags": {
    "latest": "0.0.0-dev"
  },
  "_uplinks": {},
  "_distfiles": {},
  "_attachments": {
    "demo-0.0.0-dev.tgz": {
      "shasum": "45514c40a481596880710f08ff402682d3d6028f",
      "version": "0.0.0-dev"
    }
  },
  "_rev": "3-27d20a674f77f5eb",
  "_id": "@test/demo",
  "readme": "# This is a test package\n\nA simple package to test the registry.\n\n## Usage\nInstall it with:\n```bash\nyarn add @test/demo\n```\n\nUse it with:\n```js\nimport demo from \"@test/demo\";\n\ndemo();\n```"
}
```

### 2. I try to unpublish the package
- NPM complains about a 404 error _(but the package was there before running the command)_
```bash
npm WARN using --force I sure hope you know what you are doing.
npm ERR! code E404
npm ERR! 404 Not Found - DELETE https://MY_REGISTRY_DOMAIN/@test%2fdemo/-rev/3-009c375fe8c10158 - no such package available
npm ERR! 404 
npm ERR! 404  '@test/demo@0.0.0-dev' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
npm ERR! 404 
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/marco/.npm/_logs/2019-08-12T15_22_39_884Z-debug.log
```
- `/verdaccio_s3_db.json` is updated correctly
```json
{"list":[],"secret":"4d46495e75b0dd52555a620e08a4f46c9b72f66a77b44c32477d08d9a08eacc1"}
```
- the whole `/@test` folder has been deleted so I have no metadata to show you
- any following `npm unpublish --force` command doesn't fail as the first one _(just a warning, but `0` exitcode)_
```bash
npm WARN using --force I sure hope you know what you are doing.
- undefined
```
### 3. The same goes for re-publishing
- I publish a package for the first time: _everything fine_
- try to re-publish it with `--force`: _404 error, but all files are updated/deleted correctly as in **step 2**_
- try to re-publish again: _everything fine_

### Side-note
Since **force-pushing** twice was working _(not optimal, but working)_ I decided to test the deployment on my cluster. This showed that my `Dockerfile` was bad (maybe macOS leaves much more freedom to mess up file permissions, but containers running on Linux were failing at boot). By the way I fixed it and now it boots correctly, but the issue is still there.

**Comment by flyfishMT** [comment]
> 
> 
> Did you see the lates release ? Previous version of verdaccio 4 might have corrupted the metadata. Try again with the latest version or you might need to update it manually. The problem is fixed in 4.2.1
> 
> One thing to be clear, the latest update might be problematic since does not autofix the corrupted metadata. I can explain how to do it manually is really straightforward.

Hi, I'm running a server installation that is v4.0.1 and I'm not able to upgrade at this point.  I removed a version of a package and it is still showing on this site, and I can't publish this version again so I think I have this problem. Can you explain how to fix the corrupted metadata? Thanks!

**Comment by liarco** [comment]
For your information I’m still having the issue with `4.3.3` so I don’t think that upgrading would be a solution.

I’ve never had the need to remove something, but just to replace it and I succeeded by sending the command twice in a row... I don’t know if it helps in your case.

**Comment by juanpicado** [comment]
@flyfishMT 
> Can you explain how to fix the corrupted metadata? Thanks!

The metadata is just a JSON file, quite simple to read, if you want to get rid of one specific version, you need to remove it from different sections.
![Json_Parser_Online](https://user-images.githubusercontent.com/558752/68504373-1c6bca00-0265-11ea-83a5-0f6af520156b.png)

This is something that handles the client, the fields added by verdaccio are prefixed with underscore eg: `_distfiles`.

If you don't have published packages in the X package, then just do as @liarco suggested, remove the package.json file or the whole folder.

**Comment by favoyang** [comment]
I test with 4.3.4 + s3. The published package has only one version.

unpublish with error
```
$ npm unpublish com.bastianblokland.enumgenerator --force
npm WARN using --force I sure hope you know what you are doing.
npm ERR! code E404
npm ERR! 404 Not Found - DELETE https://package.openupm.com/com.bastianblokland.enumgenerator/-rev/3-96bdb9588dc790a8 - no such package available
npm ERR! 404
npm ERR! 404  'com.bastianblokland.enumgenerator@latest' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
npm ERR! 404
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\mac\AppData\Roaming\npm-cache\_logs\2019-11-15T15_58_40_023Z-debug.log
```

Logs
```
1|verdaccio  |  info <-- 127.0.0.1 requested 'DELETE /com.bastianblokland.enumgenerator/-rev/3-96bdb9588dc790a8'
1|verdaccio  |  http <-- 404, user: openupm(61.51.132.196 via 127.0.0.1), req: 'DELETE /com.bastianblokland.enumgenerator/-rev/3-96bdb9588dc790a8', error: no such package available
```

However the package has been deleted, verified from both s3, rest api and the web portal.

**Comment by liarco** [comment]
@favoyang it seems like you are the first able to reproduce my issue, isn’t it the same behavior as https://github.com/verdaccio/verdaccio/issues/1435#issuecomment-520477460?

**Comment by favoyang** [comment]
@liarco looks the same.

**Comment by favoyang** [comment]
@liarco I have verified the issue only appears in S3, and fixed in https://github.com/verdaccio/monorepo/pull/289, see the PR description for details.

**Comment by liarco** [comment]
That’s awesome, I hope it will be merged and released soon! 😃

**Comment by juanpicado** [comment]
We will check next week, sorry guys I am super busy these days.

**Comment by favoyang** [comment]
I re-visited the issue today, that unfortuantely I only fixed `unpublish -f` in https://github.com/verdaccio/monorepo/pull/289, but `unpublish version`, `publish -f`, are also buggy. Will look into it later.

**Comment by favoyang** [comment]
Here's the breakdown for all bugs mentioned in the issue.

TLDR: jump to the last paragraph.

Bug 1 - `npm unpublish -f` failed with error, but package is actually removed by verdaccio.
- storage: s3
- reason: verdaccio returns 404 error when removing package folder at the last step, which is harmless for s3 storage. Ignore the error fixed the issue.
- fix: https://github.com/verdaccio/monorepo/pull/289

Fact 2 - `npm unpublish [-f] pkg@version` work as expect after patching PR https://github.com/verdaccio/monorepo/pull/289

Bug 3 - `npm publish -f` (republish) failed with error, but the package@version file is removed by verdaccio (`npm publish -f` == `npm unpublish -f`). Thus run the publish command again will succeed, because verdaccio thinks you're publishing a new version.
- storage: **all**
- registry.npm.org behaviors
  - republish same version is not allowed, returns error 403 - "Cannot publish over previously published version"
  - unpublish pkg@version is allowed. However, publish just unpublished pkg@version is not allowed, returns error 403. That means you can not use this trick to overwrite a published version. This is not considered as a bug, because it is consistent as above.
  - the only way to fix issues in a published version is to publish a new version instead.
- reason 
  - when work with registry.npm.org
    - the npm client requests `publish -f`, the server returns a 403 (permission error)
    - the npm client stops here
    ```
    npm http fetch PUT 403 https://registry.npmjs.org/com.bastianblokland.enumgenerator 6995ms
    npm verb stack Error: 403 Forbidden - PUT https://registry.npmjs.org/com.bastianblokland.enumgenerator - You cannot publish over the previously published versions: 1.6.1.
    npm verb stack     at C:\Program Files\nodejs\node_modules\npm\node_modules\npm-registry-fetch\check-response.js:104:15
    npm verb stack     at processTicksAndRejections (internal/process/task_queues.js:93:5)
    npm verb statusCode 403
    ```
  - when work with verdaccio,
    - the npm client requests `publish -f`, the server returns 409 (conflict error)
    - the npm client continue to work because of `-f`, fetches package info, the server returns 200
    - the npm client fetches package info again, the server returns 304 (no change)
    - the npm client requests to delete the pkg@version, the server returns 201 (successfully deleted)
    - the npm client raises version conflict error. **Here the issue, the npm client shall skip the version check, because of `-f`.**
    - this end up with the result that pkg@version being removed, with no pkg@version get published. (aka `publish -f pkg@version` == `unpubish pkg@version`)
    ```
    npm http fetch PUT 409 http://127.0.0.1:4873/com.bastianblokland.enumgenerator 78ms
    npm http fetch GET 200 http://127.0.0.1:4873/com.bastianblokland.enumgenerator?write=true 30ms
    npm WARN publish Forced publish over com.bastianblokland.enumgenerator@1.6.1
    npm http fetch GET 304 http://127.0.0.1:4873/com.bastianblokland.enumgenerator?write=true 35ms (from cache)
    npm http fetch DELETE 201 http://127.0.0.1:4873/com.bastianblokland.enumgenerator/-rev/3-3da368dddcba208e 13ms
    npm verb stack Error: Cannot publish com.bastianblokland.enumgenerator@undefined over existing version.
    npm verb stack     at ConflictError (C:\Program Files\nodejs\node_modules\npm\node_modules\libnpmpublish\publish.js:211:24)
    npm verb stack     at patchMetadata (C:\Program Files\nodejs\node_modules\npm\node_modules\libnpmpublish\publish.js:164:11)
    npm verb stack     at C:\Program Files\nodejs\node_modules\npm\node_modules\libnpmpublish\publish.js:63:22
    npm verb stack     at tryCatcher (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\util.js:16:23)
    npm verb stack     at Promise._settlePromiseFromHandler (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\promise.js:517:31)
    npm verb stack     at Promise._settlePromise (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\promise.js:574:18)
    npm verb stack     at Promise._settlePromise0 (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\promise.js:619:10)
    npm verb stack     at Promise._settlePromises (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\promise.js:699:18)
    npm verb stack     at _drainQueueStep (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\async.js:138:12)
    npm verb stack     at _drainQueue (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\async.js:131:9)
    npm verb stack     at Async._drainQueues (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\async.js:147:5)
    npm verb stack     at Immediate.Async.drainQueues [as _onImmediate] (C:\Program Files\nodejs\node_modules\npm\node_modules\bluebird\js\release\async.js:17:14)
    npm verb stack     at processImmediate (internal/timers.js:439:21)
    npm verb pkgid com.bastianblokland.enumgenerator
    npm verb cwd C:\Users\favoy\Documents\projects\verdaccio\enum-generator-unity
    npm verb Windows_NT 10.0.18363
    npm verb argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "--registry" "http://127.0.0.1:4873" "publish" "-f" "--verbose"
    npm verb node v12.13.1
    npm verb npm  v6.12.1
    npm ERR! code EPUBLISHCONFLICT
    npm ERR! publish fail Cannot publish over existing version.
    npm ERR! publish fail Update the 'version' field in package.json and try again.
    npm ERR! publish fail
    npm ERR! publish fail To automatically increment version numbers, see:
    npm ERR! publish fail     npm help version
    npm verb exit [ 1, true ]
    ```
  - **fix: won't fix. It is not a verdaccio server issue, but a npm client issue.**

Based on the analysis
- I suggest verdaccio merge https://github.com/verdaccio/monorepo/pull/289 to fix `npm unpublish -f` bug for s3 storage.
- @liarco because of the `publish -f` bug is not fixable by verdaccio. I suggest you run `npm unpublish pkg@version`, then `npm publish` to simulate the `npm publish -f`.

**Comment by liarco** [comment]
@favoyang thank you so much for your detailed explanation (I really appreciate that).
I perfectly get the point and your suggestions make absolutely sense.

**Comment by juanpicado** [comment]
@liarco let us know whether all is ok when you have some time, thanks :-)

**Comment by liarco** [comment]
Hi @juanpicado, I left a feedback yesterday about this: https://github.com/verdaccio/monorepo/pull/289#issuecomment-569282128

There’s still the problem with `publish -f` but it cannot be fixed (as @favoyang explained), by the way the 404 error on unpublish si gone!

Thank you! 😊
